src/librustc_mir/build/mod.rs

```diff
@@ -84,83 +84,100 @@ pub fn mir_build<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> Mir<'t
-                }
-                _ => None,
-            };
-
-            // FIXME: safety in closures
-            let safety = match fn_sig.unsafety {
-                hir::Unsafety::Normal => Safety::Safe,
-                hir::Unsafety::Unsafe => Safety::FnUnsafe,
-            };
-
-            let body = tcx.hir.body(body_id);
-            let explicit_arguments =
-                body.arguments
-                    .iter()
-                    .enumerate()
-                    .map(|(index, arg)| {
-                        let owner_id = tcx.hir.body_owner(body_id);
-                        let opt_ty_info;
-                        let self_arg;
-                        if let Some(ref fn_decl) = tcx.hir.fn_decl(owner_id) {
-                            let ty_hir_id = fn_decl.inputs[index].hir_id;
-                            let ty_span = tcx.hir.span(tcx.hir.hir_to_node_id(ty_hir_id));
-                            opt_ty_info = Some(ty_span);
-                            self_arg = if index == 0 && fn_decl.implicit_self.has_implicit_self() {
-                                match fn_decl.implicit_self {
-                                    hir::ImplicitSelfKind::Imm => Some(ImplicitSelfKind::Imm),
-                                    hir::ImplicitSelfKind::Mut => Some(ImplicitSelfKind::Mut),
-                                    hir::ImplicitSelfKind::ImmRef => Some(ImplicitSelfKind::ImmRef),
-                                    hir::ImplicitSelfKind::MutRef => Some(ImplicitSelfKind::MutRef),
-                                    _ => None,
-                                }
-                            } else {
-                                None
-                            };
-                        } else {
-                            opt_ty_info = None;
-                            self_arg = None;
-                        }
-                        ArgInfo(fn_sig.inputs()[index], opt_ty_info, Some(&*arg.pat), self_arg)
-                    });
-
-            let arguments = implicit_argument.into_iter().chain(explicit_arguments);
-
-            let (yield_ty, return_ty) = if body.is_generator {
-                let gen_sig = match ty.sty {
-                    ty::Generator(gen_def_id, gen_substs, ..) =>
-                        gen_substs.sig(gen_def_id, tcx),
-                    _ =>
-                        span_bug!(tcx.hir.span(id), "generator w/o generator type: {:?}", ty),
-                };
-                (Some(gen_sig.yield_ty), gen_sig.return_ty)
-            } else {
-                (None, fn_sig.output())
-            };
-
-            build::construct_fn(cx, id, arguments, safety, abi,
-                                return_ty, yield_ty, return_ty_span, body)
         } else {
-            build::construct_const(cx, body_id, return_ty_span)
+            match cx.body_owner_kind {
+                hir::BodyOwnerKind::Fn | hir::BodyOwnerKind::Closure => {
```

Comment:
You could reduce the diff and rightwards drift here (and above) my using `if let hir::BodyOwnerKind::Fn | hir::BodyOwnerKind::Closure = ...`.

Reply:
I don't think that's possible. Or is this a new feature?